### PR TITLE
Allow `DataParallel` and `jit.trace`

### PR DESCRIPTION
**Pull request by jatentaki**
1. Changed some tensors which used to be updated with in-place sliced mutation to instead concatenate lists of temporary results. This may have resulted in some increase of memory footprint, but the previous code was technically UB from the perspective of autograd (as far as I understand the documentation and pytorchers on slack), so now it plays nicely with `nn.DataParallel` and `jit.trace`.
2. Reorganized the tests, moved to top-level `tests` directory. The `all` runner script is ugly, but I couldn't get `python -m unittest discover` to work. Also note that `all` runs all tests in the directory, not all in the whole project
3. Lowered the memory footprint of batchnorm slightly by detaching a variable earlier.

Hopefully the commit messages are more descriptive of particular modifications. I tried to test this under `0.4.1` and `1.0 preview`, but please test this yourself before merging

### CODE REVIEW

**Review by mariogeiger — CHANGES_REQUESTED** [review]
Be careful, `field_mean` and `field_norm` are used later to normalize the fields.
We need the gradient to go trough them !

I think this modification is not good.

File: se3cnn/kernel.py [hunk 1]
```diff
@@ -373,13 +375,19 @@ def combination(self, weight):
                     weight_index += mi * mj * b_el
 
                     ker = torch.einsum("uvb,bijxyz->uivjxyz", (w, kij)).contiguous()  # [u, i, v, j, x, y, z]
-                    kernel[si, sj] = ker.view_as(kernel[si, sj])
                 else:
-                    kernel[si, sj] = 0
+                    ker = torch.tensor(0.)
```
Comment (mariogeiger) [inline se3cnn/kernel.py]: Please always use things like `weight.new_tensor` to ensure have the same `device` and `dtype`.
Then I think you cannot apply `view` on it. At least `torch.tensor(0.).view(2, 2)` crash on my computer.

**Review by mariogeiger — CHANGES_REQUESTED** [review]

**Review by mariogeiger — CHANGES_REQUESTED** [review]
Looks super cool. If jit works it's amazing.

Please test also the gradients. Please compare your implementation output+gradient with the current version. They must match.

File: se3cnn/kernel.py (hunk 1 above)
Comment (mariogeiger) [reply se3cnn/kernel.py]: You could add a test that has at least one `kij is None`.
`se3cnn.kernel.SE3Kernel([(1, 0)], [(1, 0), (1, 5)], 3)` does the trick

**Comment by mariogeiger** [comment]
I see that some code with `dyn_iso` is removed. Is it because you don't have the last commits from master ? I don't know if I have to worry about that or if it will be managed during merge.

**Comment by jatentaki** [comment]
I addressed most issues
1) The gradient indeed needs to flow there, fixed. Additionally, I added gradtests for both `SE3BatchNorm` and `SEBNConv`. The former passes, the latter fails due to what appears to numerical errors. Note that it also fails with the original implementation (current master). Do you have any idea how to address that?
2) Can you think of a better way to test correctness than to compare with the old version? It feels awkward to keep both for this purpose, plus realistically the older version is also not guaranteed to be bug-free. I would rather trust some actual sanity checks. Checking for equivariance follows this line of thought, but I agree more tests would be useful.
3) Fixed the `kij is None` issue. I did not use `new_tensor` but `torch.zeros(..., device=old.device, dtype=old.dtype)`, which effectively does the same thing.
4) I have no what this `dyn_iso` thing is, I have never touched it. I am trying to figure out how to pull your upstream changes to my fork.

File: tests/batchnorm_tests.py [hunk 2]
```diff
@@ -54,6 +54,19 @@ def unrotate(t):
...
+        Rs_in = [(1, 0), (1, 1)]
+        Rs_out = [(1, 0)]
+        size = 5
+
+        conv = SE3BNConvolution(Rs_in, Rs_out, size).type(torch.float64)
+
+        x = torch.rand(1, sum(m * (2 * l + 1) for m, l in Rs_in), 6, 6, 6,
+                       requires_grad=True, dtype=torch.float64)
+
+        self.assertTrue(torch.autograd.gradcheck(conv, (x, ), eps=1))
```
Comment (mariogeiger) [inline tests/batchnorm_tests.py]: It is not a linear function so it should not work with `eps=1`, did you try with a smaller value ?

File: tests/batchnorm_tests.py [hunk 3]
```diff
@@ -54,6 +54,19 @@ def unrotate(t):
...
+        mult_dim = [(1, 1), (1, 3)]
+
+        size = 5
+
+        bn = SE3BatchNorm(mult_dim, size).type(torch.float64)
+
+        x = torch.rand(1, 4, 6, 6, 6,
+                       requires_grad=True, dtype=torch.float64)
+
+        self.assertTrue(torch.autograd.gradcheck(bn, (x, ), eps=1))
```
Comment (mariogeiger) [inline tests/batchnorm_tests.py]: Same here, `y = x / sqrt(var(x))` is nonlinear therefore `eps` should be small

**Review by mariogeiger — COMMENTED** [review]
If you want to check the derivative of BN you can also check with a very simple test vs hand calculus:
```python
x = torch.rand(3, requires_grad=True).view(1, 3, 1, 1, 1)
bn = se3cnn.SE3BatchNorm([(1, 3)], eps=0)
y = bn(x)

# by hand dyi/dxj = delta_ij / |x| - xi xj / |x|^3

jac_num = jacobian(y, x)  # github.com/mariogeiger/hessian
jac_cal = torch.eye(3) / x.norm() - x.view(-1, 1) * x.view(1, -1) / x.norm() ** 3

assert jac_cal == jac_num
```

For simplicity I took `eps=0` but one could also consider it into the calculus as well.
